src/forest.rs

```diff
@@ -304,299 +304,537 @@ impl<'i, G: GrammarReflector, I: Input> ParseForest<'i, G, I> {
-        type Item = B::Output;
-        fn next(&mut self) -> Option<Self::Item> {
-            loop {
-                if let Some(ref mut b_iter) = self.b_iter {
-                    if let x @ Some(_) = b_iter.next() {
-                        return x;
-                    }
-                }
-                match self.a_iter.next() {
-                    // HACK(eddyb) this never does anything, but without a *second*
-                    // call to `B::Iter::next`, LLVM spends more time optimizing.
-                    None => {
-                        return match self.b_iter_backwards {
-                            Some(ref mut b_iter) => b_iter.next(),
-                            None => None,
-                        }
-                    }
-                    Some(x) => self.b_iter = Some(self.b_arrow.apply(x)),
+    pub struct Split<Left: Shape, Right: Shape>(PhantomVoid<(Left, Right)>);
+
+    impl<Left: Shape, Right: Shape> Shape for Split<Left, Right> {
+        const STATE_LEN: usize = 1 + Left::STATE_LEN + Right::STATE_LEN;
+
+        fn init<'i, G: GrammarReflector, I: Input>(
+            forest: &ParseForest<'i, G, I>,
+            node: Node<'i, G>,
+            state: &mut [usize],
+        ) {
+            let (state_split, state) = state.split_at_mut(1);
+            let state_split = &mut state_split[0];
+            let (state_left, state_right) = state.split_at_mut(Left::STATE_LEN);
+
+            let &split = forest.possibilities[&node].iter().next().unwrap();
+
+            let (left, right) = forest.split_children(node, split);
+
+            *state_split = split;
+            Left::init(forest, left, state_left);
+            Right::init(forest, right, state_right);
+        }
+        fn read<'i, G: GrammarReflector, I: Input>(
+            forest: &ParseForest<'i, G, I>,
+            node: Node<'i, G>,
+            state: &[usize],
+            fields: &mut [Option<Node<'i, G>>],
+        ) {
+            let (state_split, state) = state.split_at(1);
+            let state_split = state_split[0];
+            let (state_left, state_right) = state.split_at(Left::STATE_LEN);
+
+            let (left, right) = forest.split_children(node, state_split);
+            Left::read(forest, left, state_left, fields);
+            Right::read(forest, right, state_right, fields);
+        }
+        fn step<'i, G: GrammarReflector, I: Input>(
+            forest: &ParseForest<'i, G, I>,
+            node: Node<'i, G>,
+            state: &mut [usize],
+        ) -> bool {
+            let (state_split, state) = state.split_at_mut(1);
+            let state_split = &mut state_split[0];
+            let (state_left, state_right) = state.split_at_mut(Left::STATE_LEN);
+
+            let (left, right) = forest.split_children(node, *state_split);
+
+            Right::step(forest, right, state_right)
+                || Left::step(forest, left, state_left) && {
```

Comment:
Offtopic: You know, I don't really agree with rustfmt's choice here, though I don't know what would be better. It makes me think closure. Maybe that's on Rust for using `||` for both logical OR and the empty closure header, though.

Reply:
Or maybe I can find a better way to write this, idk. I basically want to bail on success instead of error, maybe I should use a `Result<Finished, Stepped>` but it kind of hurts to think about.

Reply:
`Option<Finished>` might make more sense.